File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void inorder(TreeNode root,List<Integer> ls){
       if(root==null) return;
        inorder(root.left,ls);
        ls.add(root.val);
        inorder(root.right,ls);
    }
    public boolean isValidBST(TreeNode root) {
        List<Integer> ls = new ArrayList<>();
        inorder(root,ls);
        for(int i=1;i<ls.size();i++){
            if(ls.get(i)<=ls.get(i-1)) return false;
        }   
        return true;
    }
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
struct ReturnType {
    bool res;
    long max_val;
    long min_val;
};
ReturnType soln(TreeNode* root) {
    if (!root) {
        return {true, LONG_MIN, LONG_MAX};
    }
    ReturnType l = soln(root->left);
    ReturnType r = soln(root->right);
    bool ans = l.res && r.res && root->val > l.max_val && root->val < r.min_val;
    return {ans, max((long)root->val, r.max_val),
            min((long)root->val, l.min_val)};
}
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        ReturnType ans = soln(root);
        return ans.res;
    }
};